pkg/virtual/apiexport/authorizer/maximal_permission_policy.go

```diff
@@ -0,0 +1,129 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package authorizer
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kaudit "k8s.io/apiserver/pkg/audit"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/client-go/kubernetes"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	kcpauth "github.com/kcp-dev/kcp/pkg/authorization"
+	"github.com/kcp-dev/kcp/pkg/authorization/delegated"
+	kcpclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	dynamiccontext "github.com/kcp-dev/kcp/pkg/virtual/framework/dynamic/context"
+)
+
+type maximalPermissionAuthorizer struct {
+	getAPIExport           func(ctx context.Context, clusterName, apiExportName string) (*apisv1alpha1.APIExport, error)
+	newDelegatedAuthorizer func(clusterName string) (authorizer.Authorizer, error)
+}
+
+func NewMaximalPermissionAuthorizer(deepSARClient kubernetes.ClusterInterface, kcpClusterClient kcpclient.ClusterInterface) authorizer.Authorizer {
+	return &maximalPermissionAuthorizer{
+		getAPIExport: func(ctx context.Context, clusterName, apiExportName string) (*apisv1alpha1.APIExport, error) {
+			return kcpClusterClient.Cluster(logicalcluster.New(clusterName)).ApisV1alpha1().APIExports().Get(ctx, apiExportName, v1.GetOptions{})
+		},
+		newDelegatedAuthorizer: func(clusterName string) (authorizer.Authorizer, error) {
+			return delegated.NewDelegatedAuthorizer(logicalcluster.New(clusterName), deepSARClient)
+		},
+	}
+}
+
+func (a *maximalPermissionAuthorizer) Authorize(ctx context.Context, attr authorizer.Attributes) (dec authorizer.Decision, reason string, err error) {
+	defer func() {
+		auditReason := reason
+		if err != nil {
+			auditReason = fmt.Sprintf("%v, error: %v", reason, err)
+		}
+		kaudit.AddAuditAnnotations(
+			ctx,
+			VirtualAPIExportMaximumPermissionPolicyAuditDecision, kcpauth.DecisionString(dec),
+			VirtualAPIExportMaximumPermissionPolicyAuditReason, auditReason,
+		)
+	}()
+
+	apiDomainKey := dynamiccontext.APIDomainKeyFrom(ctx)
+	parts := strings.Split(string(apiDomainKey), "/")
+	if len(parts) < 2 {
+		return authorizer.DecisionNoOpinion, VirtualAPIExportNotPermittedReason, fmt.Errorf("invalid API domain key: %q", apiDomainKey)
+	}
+	apiExportCluster := parts[0]
+	apiExportName := parts[1]
+
+	apiExport, err := a.getAPIExport(ctx, apiExportCluster, apiExportName)
```

Comment:
is this the right APIExport ? The domain key points to the red APIExport B in https://docs.google.com/drawings/d/1_sOiFZReAfypuUDyHS9rwpxbZgJNJuxdvbgXXgu2KAQ/edit, doesn't it?

Reply:
yes, indeed 😱 i have to rework this part.